HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs

```diff
@@ -0,0 +1,172 @@
+        {
+            _cache = cache;
+            _username = username;
+            _password = password;
+            _httpHost = httpHost;
+            _urlBase = $"https://{_httpHost}/api";
+
+            return await AuthenticateAsync();
+        }
+
+        /// <summary>
+        /// Authenticates and runs the URL with the acceptHeader
+        /// </summary>
+        /// <param name="url">Complete URL</param>
+        /// <param name="acceptHeader">header used to make the call</param>
+        /// <returns>Endpoint Result</returns>
+        public static async Task<string> RunEndpointCallAsync(string url, string acceptHeader)
+        {
+            if (_accessTokenInstance == null)
+            {
+                await AuthenticateAsync();
+            }
+
+            var result = _cache.Get(url);
+            return !string.IsNullOrEmpty(result) ? result : await RunEndpointHelperAsync(url, acceptHeader);
+        }
+
+        /// <summary>
+        /// Runs the URL using the required header info
+        /// </summary>
+        /// <param name="url">Complete URL</param>
+        /// <param name="acceptHeader">header used to make the call</param>
+        /// <returns>EndPoint result</returns>
+        private static async Task<string> RunEndpointHelperAsync(string url, string acceptHeader)
+        {
+            string rawJsonToReturn = string.Empty;
+            var uri = new Uri(url);
+
+            using (var handler = new HttpClientHandler
+            {
+                ServerCertificateCustomValidationCallback = (sender, certificate, chain, sslPolicyErrors) => true,
+                AllowAutoRedirect = false
+            })
+            using (var httpClient = new HttpClient(handler))
+            {
+                httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _accessTokenInstance.Access_Token);
+                var response = await httpClient.GetAsync(uri);
+                
+                if (response.StatusCode == HttpStatusCode.Moved || response.StatusCode == HttpStatusCode.Redirect)
+                {
+                    return await RunEndpointHelperAsync(response.Headers.Location.ToString(), acceptHeader);
+                }
+                else
+                {
+                    rawJsonToReturn = await response.Content.ReadAsStringAsync();
+                }
+            }
+
+            _cache.Add(url, rawJsonToReturn);
+            _cache.Persist();
+
+            return rawJsonToReturn ?? string.Empty;
+        }
+
+        /// <summary>
+        /// Authenticate access using the host location
+        /// </summary>
+        /// <returns>bool: True = success</returns>
+        private static async Task<bool> AuthenticateAsync()
+        {
+            var body = $"{{\"username\": \"{_username}\",\"password\": \"{_password}\"}}";
+
+            var uri = new Uri($"{UrlBase}/api/authentication/login");
```

Comment:
wrong api endpoint

Reply:
see documentation

Reply:
the external route was ready when this is coded. will update.